lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -34,420 +25,540 @@
+          if (reachingDefsToSrcPort != modulePaths->second.end())
+            if (llvm::find(reachingDefsToSrcPort->second, modSinkPortField) !=
+                reachingDefsToSrcPort->second.end()) {
+              // This case can occur when there are multiple RWProbes on the
+              // port, which refer to the same base value. So, each of such
+              // probes are reaching defs of each other. Hence the false
+              // loops. Instead of recording this in the reachingDefGraph,
+              // record it separately with the rwProbeClasses.
+              setOfEquivalentRWProbes.insert(srcNode);
+              if (minArgNum > srcArgNum) {
+                // Make one of the RWProbe port the base node. Use the first
+                // port for deterministic error messages.
+                minArgNum = srcArgNum;
+                basePortNode = srcNode;
+              }
+              continue;
+            }
         }
-        FieldRef instanceOutPort(inst.getResult(outPortNum),
-                                 modOutPort.getFieldID());
-        llvm::append_range(children, fieldToVals[instanceOutPort]);
+        addReachingDef(sinkPort, srcPort);
       }
-    } else if (auto mem = dyn_cast<MemOp>(ref.getDefiningOp())) {
-      if (mem.getReadLatency() > 0)
+      if (setOfEquivalentRWProbes.empty())
+        continue;
+
+      // Add all the rwprobes to the same class.
+      for (auto probe : setOfEquivalentRWProbes)
+        rwProbeClasses.unionSets(probe, sinkNode);
+
+      // Make the first port as the base value.
+      // Note: this is a port and the actual reference base exists in another
+      // module.
+      auto leader = rwProbeClasses.getLeaderValue(sinkNode);
+      rwProbeRefersTo[leader] = basePortNode;
+
+      setOfEquivalentRWProbes.insert(sinkNode);
+      // Add a reaching def  from base RWProbe port to all other RWProbe ports.
+      for (auto probe : setOfEquivalentRWProbes)
+        if (probe != basePortNode)
+          reachingDefGraph[probe].second.push_back(basePortNode);
+    }
+  }
+
+  // Record the FieldRef, corresponding to the result of the sub op
+  // `result = base[index]`
+  void recordValueRefersToFieldRef(Value base, unsigned fieldID, Value result) {
+
+    // Check if base is itself field of an aggregate.
+    auto it = valToFieldRefs.find(base);
+    if (it != valToFieldRefs.end()) {
+      // Rebase it to the original aggregate.
+      // Because of subaccess op, each value can refer to multiple FieldRefs.
+      SmallVector<FieldRef> entry;
+      for (auto &sub : it->second)
+        entry.emplace_back(sub.getValue(), sub.getFieldID() + fieldID);
+      // Update the map at the end, to avoid invaliding the iterator.
+      valToFieldRefs[result].append(entry.begin(), entry.end());
+      return;
+    }
+    // Break cycles from registers.
+    if (auto *def = base.getDefiningOp()) {
+      if (isa<RegOp, RegResetOp, SubfieldOp, SubindexOp, SubaccessOp>(def))
```

Comment:
Why can we ignore subelement ops? I think we need to strip subelements ops beforehand like `if (auto * def = getFieldRefFromValue(base).getValue().getDefiningOp())`

Reply:
I think you mean this condition as an assert instead of the return, 
this condition is not an assert, this is expected in the IR, and this is just a "sub"  with a base of a register. Its required to break the non-comb loops.